[PyTorch Mobile] Enable partial loading of GPU models on linux CPU machines

dhruvbird opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51236 [PyTorch Mobile] Enable partial loading of GPU models on linux CPU machines**

The problem we currently have with tracing is that GPU models can't load on devvm CPU machines. Here's why:

1. The Metal GPU ops don't exist so the validation that checks for missing ops kicks in and prevents loading
2. Even if the check for missing ops is commented out, the actual model contents can't be succssfully loaded (see t83364623 for details)

Hence, to work around these problems and allow tracing to detect GPU models, and skip actual tracing for these (as discussed in the meeting 2 weeks ago and based on recommendations from @razy, @myuan, and @taox), we're adding code to detect these GPU models based on the set of operators that show up in the file `extra/mobile_info.json`.

The code then skips tracing, and picks up the root operators from the model itself.

The diff below this one will be removed before landing since we don't want to check in the model - I've kept it here in case anyone wants to patch this diff in and run the command on their devvm locally.

Differential Revision: [D26109526](https://our.internmc.facebook.com/intern/diff/D26109526/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26109526/)!

dhruvbird committed:
[PyTorch Mobile] Enable partial loading of GPU models on linux CPU machines

The problem we currently have with tracing is that GPU models can't load on devvm CPU machines. Here's why:

1. The Metal GPU ops don't exist so the validation that checks for missing ops kicks in and prevents loading
2. Even if the check for missing ops is commented out, the actual model contents can't be succssfully loaded (see t83364623 for details)

Hence, to work around these problems and allow tracing to detect GPU models, and skip actual tracing for these (as discussed in the meeting 2 weeks ago and based on recommendations from @razy, @myuan, and @taox), we're adding code to detect these GPU models based on the set of operators that show up in the file `extra/mobile_info.json`.

The code then skips tracing, and picks up the root operators from the model itself.

The diff below this one will be removed before landing since we don't want to check in the model - I've kept it here in case anyone wants to patch this diff in and run the command on their devvm locally.

Differential Revision: [D26109526](https://our.internmc.facebook.com/intern/diff/D26109526/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26109526/)!

[ghstack-poisoned]

ljk53 commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -509,5 +518,20 @@ mobile::Module _load_for_mobile(
+  auto observer = torch::observerConfig().getModuleObserver();
+  auto instance_key = std::rand();
+  if (observer) {
+    observer->onEnterLoadModel(instance_key);
```
nit: shall we fire the same onEnterLoadModel event if this is only to load extra meta without intent to run the inference? (would this affect the metric readings?)

dhruvbird committed:
Update on "[PyTorch Mobile] Enable partial loading of GPU models on linux CPU machines"


The problem we currently have with tracing is that GPU models can't load on devvm CPU machines. Here's why:

1. The Metal GPU ops don't exist so the validation that checks for missing ops kicks in and prevents loading
2. Even if the check for missing ops is commented out, the actual model contents can't be succssfully loaded (see t83364623 for details)

Hence, to work around these problems and allow tracing to detect GPU models, and skip actual tracing for these (as discussed in the meeting 2 weeks ago and based on recommendations from @razy, @myuan, and @taox), we're adding code to detect these GPU models based on the set of operators that show up in the file `extra/mobile_info.json`.

The code then skips tracing, and picks up the root operators from the model itself.

The diff below this one will be removed before landing since we don't want to check in the model - I've kept it here in case anyone wants to patch this diff in and run the command on their devvm locally.

Differential Revision: [D26109526](https://our.internmc.facebook.com/intern/diff/D26109526/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26109526/)!

[ghstack-poisoned]
